operations/research-and-development/product/development-process/secure-software-development.md

```diff
@@ -0,0 +1,77 @@
+
+## Recommended reading
+
+* [OSSF Scorecard Action Check Documentation](https://github.com/ossf/scorecard/blob/main/docs/checks.md)
+* [Overview of Software Supply Chain Attacks](https://www.crowdstrike.com/cybersecurity-101/cyberattacks/supply-chain-attacks/) by [Crowdstrike](https://www.crowdstrike.com/)
+
+## Code reviews
+
+### Manual code reviews
+
+The product security team performs reviews on new features and plugins when necessary. Here's some of the things included in the checks:
+
+* TODO.
+
+### Automated code reviews
+
+The key Mattermost repositories have (or should have) automated code security analysis tools set up. In practice, we use [CodeQL](https://codeql.github.com/) as a Github Action to automatically analyze all pull requests. Additionally, the use of the [OSSF Security Scorecards GitHub Action](https://github.com/ossf/scorecard) is strongly encouraged. This document addresses many of the best practices checked by the Scorecards action.
+
+## Dependency pinning
+
+Dependency pinning hardens the software development workflow against [software supply chain attacks](https://www.crowdstrike.com/cybersecurity-101/cyberattacks/supply-chain-attacks/). In this context, the specific threat being mitigated is upstream dependencies being taken over by attackers and replaced with malicious content. The risk commonly involves referring to dependencies without explicitly specifying a computed hash value for the dependency. In case of a dependency takeover, pulling the _latest_ version of it realizes the risk. The risk probability is considered low for supply chain attacks, but the effect can be pretty bad.
+
+Practically, defining build-time dependencies with an explicit hash is the preferred solution. When specifying hashes to existing dependencies that did not use hashed references before, picking the hash of the currently used dependency is a good starting point as it is often not practical to do a full audit of all dependencies. Picking a hash and sticking to it doesn't ensure that said hash represents a safe version, it only ensures that a possibly malicious later version is not introduced in the build process afterwards.
+
+Please note also that pinning dependencies to a specific version (for example, `@2.1.1`) does not guard against malicious updates to the dependency. It is necessary to pin a dependency to a specific checksum (which is a cryptograhic guarantee to the uniqueness of the actual content).
+
+### Npm dependencies
+
+With npm, replace `npm install` with `npm ci` everywhere in the build/CI pipeline. This ensures that only packages matching `package-lock.json` are installed, see [this Stackoverflow answer](https://stackoverflow.com/questions/52499617/what-is-the-difference-between-npm-install-and-npm-ci) for more details and a reference to the npm documentation.
```

Comment:
What's the value in linking to the SO answer instead of the docs page directly? Doesn't seem like the answer adds much and the docs are more likely to stay up to date.